File: src/services/sync.ts
import { Context, Effect, Layer } from "effect"
import type { ShellExecError } from "../errors.js"
import type { ProjectConfig } from "../schema/config.js"
import { DatabaseService } from "./database.js"
import { GitService } from "./git.js"
import { ShellService } from "./shell.js"

// ---------------------------------------------------------------------------
// SyncService — Tier 3 orchestrator (D7: real `layer` only, no layerMemory).
// Fast-forwards the project's base checkout, then (only when HEAD moved) runs
// install/generate and — when the database is reachable (ping) — migrate.
// ---------------------------------------------------------------------------

export interface SyncResult {
  readonly fetched: boolean
  readonly pulled: boolean
  readonly headMoved: boolean
  readonly installed: boolean
  readonly migrated: boolean
  readonly skippedPull?: string
}

export interface SyncShape {
  readonly sync: (
    config: ProjectConfig,
    base?: string
  ) => Effect.Effect<SyncResult, ShellExecError>
}

export class SyncService extends Context.Tag("ship/SyncService")<
  SyncService,
  SyncShape
>() {
  static layer: Layer.Layer<
    SyncService,
    never,
    GitService | ShellService | DatabaseService
  > = Layer.effect(
    SyncService,
    Effect.gen(function* () {
      const git = yield* GitService
      const shell = yield* ShellService
      const db = yield* DatabaseService

      const sync: SyncShape["sync"] = Effect.fn("SyncService.sync")(
        function* (config, baseBranch) {
          const repoPath = config.path

          // 1. Fetch origin.
          yield* git.fetch(repoPath)

          // 2a. Custom base: fast-forward that branch ref directly (no checkout).
          if (baseBranch) {
            const before = yield* git.revParse(repoPath, baseBranch).pipe(
              Effect.catchTag("ShellExecError", () => Effect.succeed(""))
            )
            const updateOk = yield* git.updateBranch(repoPath, baseBranch).pipe(
              Effect.as(true),
              Effect.catchTag("ShellExecError", () => Effect.succeed(false))
            )
            if (!updateOk) {
              return {
                fetched: true,
                pulled: false,
                headMoved: false,
                installed: false,
                migrated: false,
                skippedPull: `could not fast-forward ${baseBranch}`,
              }
            }
            const after = yield* git.revParse(repoPath, baseBranch).pipe(
              Effect.catchTag("ShellExecError", () => Effect.succeed(""))
            )
            return {
              fetched: true,
              pulled: true,
              headMoved: before !== after,
              installed: false,
              migrated: false,
            }
          }

          // 2b. Default: fast-forward main (skip if dirty or non-ff).
          const dirty = yield* git.isDirty(repoPath)
          if (dirty) {
            return {
              fetched: true,
              pulled: false,
              headMoved: false,
              installed: false,
              migrated: false,
              skippedPull: "working tree has uncommitted changes",
            }
          }

          const before = yield* git.revParseHead(repoPath)
          const pullOk = yield* git.pullFfOnly(repoPath).pipe(
            Effect.as(true),
            Effect.catchTag("ShellExecError", () => Effect.succeed(false))
          )
          if (!pullOk) {
            return {
              fetched: true,
              pulled: false,
              headMoved: false,
              installed: false,
              migrated: false,
              skippedPull: "cannot fast-forward (main has diverged)",
            }
          }

          const after = yield* git.revParseHead(repoPath)
          const headMoved = before !== after

          // 3. Only install/generate/migrate when HEAD moved.
          let installed = false
          let migrated = false

          if (headMoved) {
            if (config.commands.install) {
              yield* shell.execInDir(repoPath, config.commands.install)
              installed = true
            }
            if (config.commands.generate) {
              yield* shell.execInDir(repoPath, config.commands.generate)
            }
            if (config.commands.migrate) {
              const running = yield* db.ping({
                runtime: config.database.runtime,
                user: config.database.user,
              })
              if (running) {
                yield* shell.execInDir(repoPath, config.commands.migrate)
                migrated = true
              }
            }
          }

          return { fetched: true, pulled: true, headMoved, installed, migrated }
        }
      )

      return { sync }
    })
  )
}

File: src/services/sync.rs
use crate::errors::Result;
use crate::schema::ProjectConfig;
use crate::services::database::{self, DbTarget};
use crate::services::git;
use crate::services::shell::{self, NON_INTERACTIVE_ENV};

// Fast-forwards the project's base checkout, then (only when HEAD moved) runs
// the install scope and — when the database is reachable (ping) — the db scope.

pub struct SyncResult {
    pub pulled: bool,
    pub head_moved: bool,
    pub installed: bool,
    pub migrated: bool,
    pub skipped_pull: Option<String>,
}

fn skipped(reason: String) -> SyncResult {
    SyncResult {
        pulled: false,
        head_moved: false,
        installed: false,
        migrated: false,
        skipped_pull: Some(reason),
    }
}

pub fn sync(pc: &ProjectConfig, base: Option<&str>) -> Result<SyncResult> {
    let repo = &pc.path;

    git::fetch(repo)?;

    // Custom base: fast-forward that branch ref directly (no checkout).
    if let Some(base_branch) = base {
        let before = git::rev_parse(repo, base_branch).unwrap_or_default();
        if git::update_branch(repo, base_branch).is_err() {
            return Ok(skipped(format!("could not fast-forward {base_branch}")));
        }
        let after = git::rev_parse(repo, base_branch).unwrap_or_default();
        return Ok(SyncResult {
            pulled: true,
            head_moved: before != after,
            installed: false,
            migrated: false,
            skipped_pull: None,
        });
    }

    // Default: fast-forward main (skip if dirty or non-ff).
    if git::is_dirty(repo)? {
        return Ok(skipped("working tree has uncommitted changes".to_string()));
    }

    let before = git::rev_parse_head(repo)?;
    if git::pull_ff_only(repo).is_err() {
        return Ok(skipped(
            "cannot fast-forward (main has diverged)".to_string(),
        ));
    }
    let after = git::rev_parse_head(repo)?;
    let head_moved = before != after;

    let mut installed = false;
    let mut migrated = false;

    if head_moved {
        if !pc.commands.install.is_empty() {
            for cmd in &pc.commands.install {
                shell::exec_in_dir(repo, cmd, NON_INTERACTIVE_ENV)?;
            }
            installed = true;
        }
        if !pc.commands.db.is_empty() && database::ping(DbTarget::from(&pc.database)) {
            for cmd in &pc.commands.db {
                shell::exec_in_dir(repo, cmd, NON_INTERACTIVE_ENV)?;
            }
            migrated = true;
        }
    }

    Ok(SyncResult {
        pulled: true,
        head_moved,
        installed,
        migrated,
        skipped_pull: None,
    })
}
